vignettes/developers.Rmd

```diff
@@ -0,0 +1,358 @@
+So you might be tempted to construct a sentence like this:
+
+```{r}
+cows <- function(n) {
+  if (n == 1) {
+    paste0(n, " cow") 
+  } else { 
+    paste0(n, " cows")
+  }
+}
+paste("I have ", cows(0))
+paste("I have ", cows(1))
+paste("I have ", cows(2))
+```
+
+But this doesn't always work, even in English:
+
+```{r}
+paste0("There are ", cows(0), " in the field")
+paste0("There are ", cows(1), " in the field")
+```
+
+Again, we always want to construct a complete sentence:
+
+```{r}
+field_cows <- function(n) {
+  if (n == 1) {
+    fmt <- tr_("There is {n} cow in the field")
+  } else {
+    fmt <- tr_("There are {n} cows in the field")
+  }
+  glue(fmt)
+}
+```
+
+But there's an additional wrinkle here: while English has singular and plural, other languages have different forms.
+So we need to use a different helper: `ngettext(n, singular, plural)`:
+
+```{r}
+field_cows <- function(n) {
+  glue(ngettxt(n,
+    "There is {n} cow in the field",
+    "There are {n} cows in the field"
+  ))
+}
+```
+
+`ngettext()` generates a special form in the `.po` file, which shows both singular and plural forms:
+
+    msgid "There is {n} cow in the field"
+    msgid_plural "There are {n} cows in the field"
+
+Then the translator can supply any number of translations.
+Languages that don't have plurals (e.g. Chinese) only need to supply a single translation:
+
+    msgid "There is {n} cow in the field"
+    msgid_plural "There are {n} cows in the field"
+    msgstr[0] "田裡有{n}頭牛"
+
+Russian has three forms, so it gets three entries (roughly 1, 2-4, and everything else):
+
+    msgid "There is {n} cow in the field"
+    msgid_plural "There are {n} cows in the field"
+    msgstr[0] "В поле {xn} корова"
+    msgstr[1] "В поле {n} коровы"
+    msgstr[2] "В поле {n} коров"
+
+Slovenian and Serbian have four forms, Irish has five forms (learn more at [bitesize Irish](https://www.bitesize.irish/blog/counting/)), and Arabic has six forms.
+The rules that define which number get which message are quite complex and encoded in the "plural form" that's recorded at the top of the `.po` file and looks something like `(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<10 || n%100>=20)? 1 : 2).` However, this is something the translators need to worry about not you, and as speakers of the language they should find it easier to puzzle out the rules.
+
+### Collapsed lists
+
+If you need to enumerate a list of options, you could use the [and](https://github.com/rossellhayes/and) package:
+
+```{r, eval = FALSE}
+library(and)
+values <- c("first", "middle", "last")
+or(values)
+#> [1] "first, middle, and last"
+
+# lang is normally retrieve automatically from the environemtn
+# overriding it here to show what a translation looks like:
+or(values, lang = "fr")
+#> [1] "first, middle ou last"
+```
+
+Which works will in glue:
+
+```{r, eval = FALSE}
+glue(tr_("`x` must be one of {and(values)}"))
+#> `x` must be one of first, middle and last
+```
```

Comment:
Any conclusion?

Reply:
I don't usually write conclusions in documentation, so I've got no idea what to put here. Let me know what you'd like.